docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+This will NOT check that the state exists and you can use it in case you need to upgrade contract and migrate state.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    #[init(ignore_state)]
+    pub fn migrate_state(new_data: String) -> Self {
+        // Deserialize the state using the old contract structure.
+        let old_contract: OldContract = env::state_read().expect("Old state doesn't exist");
+        // Verify that the migration can only be done by the owner.
+        // This is not necessary, if the upgrade is done internally.
+        assert_eq!(
+            &env::predecessor_account_id(),
+            &old_contract.owner_id,
+            "Can only be called by the owner"
+        );
+
+        // Create the new contract using the data from the old contract.
+        Self { owner_id: old_contract.owner_id, data: old_contract.data, new_data }
+    }
+}
+```
+
+## Use `PanicOnDefault`
+
+By default `near_sdk` allows a contract to be initialized with default state.
+Usually, if you have an initializer, you will want to prevent this.
+There is a helper derive macro `PanicOnDefault` to do this, e.g.
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    pub data: String,
+}
+```
+
+## Public vs private methods
+
+For methods in the implementation under `#[near_bindgen]`:
+
+- `pub fn` makes a method public and exports it in a contract. It means anyone can call it.
+- `fn` makes the method internal, and it's not exported from the contract. No one can call it directly. It can only be called
+  within a contract directly (not through a promise).
+- `pub(crate) fn` also will make a method internal. It's helpful to use it when you have a method in a different module.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn increment(&mut self) {
+        self.internal_increment();
+    }
+
+    fn internal_increment(&mut self) {
+        self.counter += 1;
+    }
+}
+```
+
+Another way of not exporting methods is by having a separate `impl Contract` section, that is not marked with `#[near_bindgen]`.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn increment(&mut self) {
+        self.internal_increment();
+    }
+}
+
+impl Contract {
+    /// This methods is still not exported.
+    pub fn internal_increment(&mut self) {
+        self.counter += 1;
+    }
+}
+```
+
+While this style of public and private methods are familiar from other programming contexts, smart contracts also have a more unique need: public methods (exported as part of the contract interface; callable via [cross-contract calls](https://docs.near.org/docs/tutorials/contracts/cross-contract-calls)) that are only callable by the contract itself (such as [`ft_resolve_transfer`](https://nomicon.io/Standards/Tokens/FungibleTokenCore.html#reference-level-explanation) for Fungible Token contracts).
+
+We call such methods _callbacks_.
```

Comment:
This is already covered in the `Public Method Types` section, unless there are any specific details you think should be added.

Reply:
Moved some of this to the private methods section that's worth keeping https://github.com/near/sdk-docs/pull/37/commits/7071e1c01dcf39f22c73248cf58fb28d8e9e3a78